### PR TITLE
HAWQ 459 Enhanced metadata api to support multiple items

### PR DESCRIPTION
**Pull request by shivzone**

### CODE REVIEW

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java [hunk 1]
```diff
@@ -26,51 +26,52 @@
...
+        private String path;
+        private String name;
 
-        public Table(String dbName, String tableName) {
+        public Item(String path, String itemName) {
 
-            if (StringUtils.isBlank(dbName) || StringUtils.isBlank(tableName)) {
-                throw new IllegalArgumentException("Table name cannot be empty");
+            if (StringUtils.isBlank(path) || StringUtils.isBlank(itemName)) {
+                throw new IllegalArgumentException("Item name cannot be empty");
```
Comment (GodenYao) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java]: "item or path" name?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java [hunk 2]
```diff
@@ -21,25 +26,26 @@
...
+     * @param metaData the input data
      */
-    public MetadataFetcher() {
-
+    public MetadataFetcher(InputData metaData) {
+        super(metaData);
     }
 
     /**
      * Gets a metadata of a given table
```
Comment (sansanichfb) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java]: table?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java [hunk 3]
```diff
@@ -1,5 +1,10 @@
 package org.apache.hawq.pxf.api;
 
+import java.util.List;
+
+import org.apache.hawq.pxf.api.utilities.InputData;
+import org.apache.hawq.pxf.api.utilities.Plugin;
+
```
Comment (GodenYao) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java]: It looks import lines should be after ASF header.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 4]
```diff
@@ -203,19 +205,37 @@ private static boolean verifyModifers(String[] modifiers) {
...
      *
      * @param qualifiedName Hive table name
-     * @return {@link org.apache.hawq.pxf.api.Metadata.Table} object holding the full table name
+     * @return {@link Metadata.Item} object holding the full table name
      */
-    public static Metadata.Table parseTableQualifiedName(String qualifiedName) {
+    public static Metadata.Item extractTableFromName(String qualifiedName) {
+        List<Metadata.Item> items = extractTablesFromPattern(null, qualifiedName);
+        if(items.isEmpty()) {
+            return null;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: I would rather throw an exception.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 5]
```diff
@@ -225,15 +245,53 @@ private static boolean verifyModifers(String[] modifiers) {
...
+        List<String> tables = new ArrayList<String>();
+
+        if(client == null || (!dbPattern.contains(WILDCARD) && !tablePattern.contains(WILDCARD)) ) {
+            /* This case occurs when the call is invoked as part of the fragmenter api or when metadata is requested for a specific table name */
+            itemList.add(new Metadata.Item(dbPattern, tablePattern));
+            return itemList;
+        }
+
+        try {
+            /*if(dbPattern.contains(WILDCARD)) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Remove it?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 6]
```diff
@@ -225,15 +245,53 @@ private static boolean verifyModifers(String[] modifiers) {
...
+                    }
+                }
+            }
+            if(itemList.isEmpty()) {
+                throw new IllegalArgumentException("no tables found");
+            }
+            return itemList;
+
+        } catch (MetaException cause) {
+            throw new RuntimeException("Failed connecting to Hive MetaStore service: " + cause.getMessage(), cause);
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Should we have consistent messages in expection, either starting from upper or lower letter?

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java [hunk 7]
```diff
@@ -95,6 +95,7 @@ public ProtocolData(Map<String, String> paramsMap) {
         accessor = getProperty("ACCESSOR");
         resolver = getProperty("RESOLVER");
         fragmenter = getOptionalProperty("FRAGMENTER");
+        metadata = getOptionalProperty("METADATA");
```
Comment (sansanichfb) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java]: How are we using this?

File: pxf/pxf-service/src/main/resources/pxf-profiles-default.xml [hunk 8]
```diff
@@ -49,11 +49,12 @@ under the License.
             <fragmenter>org.apache.hawq.pxf.plugins.hive.HiveDataFragmenter</fragmenter>
             <accessor>org.apache.hawq.pxf.plugins.hive.HiveAccessor</accessor>
             <resolver>org.apache.hawq.pxf.plugins.hive.HiveResolver</resolver>
+            <metadata>org.apache.hawq.pxf.plugins.hive.HiveMetadataFetcher</metadata>
         </plugins>
     </profile>
     <profile>
         <name>HiveRC</name>
-        <description>This profile is suitable only for Hive tables stored in RC files
+        <description>This profile is suitable only for Hive items stored in RC files
```
Comment (sansanichfb) [inline pxf/pxf-service/src/main/resources/pxf-profiles-default.xml]: In current context might remain "tables".

**Comment by sansanichfb** [comment]
We are introducing API change, so we should bump up pxfProtocolVersion in gradle.properties.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 9]
```diff
@@ -203,19 +205,37 @@ private static boolean verifyModifers(String[] modifiers) {
...
-        if (StringUtils.isBlank(qualifiedName)) {
+        if (StringUtils.isBlank(pattern)) {
             throw new IllegalArgumentException("empty string" + errorMsg);
         }
 
-        String[] rawToks = qualifiedName.split("[.]");
+        String[] rawToks = pattern.split("[.]");
         ArrayList<String> toks = new ArrayList<String>();
         for (String tok: rawToks) {
             if (StringUtils.isBlank(tok)) {
```
Comment (GodenYao) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: so this means input like "db_name.....table_name" will be accepted?

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java (hunk 1 above)
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java]: this checks for both

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java (hunk 2 above)
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java]: Will change to item

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java (hunk 3 above)
Comment (shivzone) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java]: you are right

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java (hunk 7 above)
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java]: This is used to when the getMetdata api is invoked. This class helps extracting the metadata class from the profile xml

**Comment by shivzone** [comment]
@sansanichfb i don't think we need to increase the API version as v14 wasn't quite consumed by any user/customer

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java (hunk 9 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Thats true but doesn't seem right. I'm not quite sure why we had this earlier

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java [hunk 10]
```diff
@@ -19,27 +19,33 @@
...
 
 /**
- * Abstract class that defines getting metadata of a table.
+ * Abstract class that defines getting metadata of an item.
```
Comment (hornn) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/MetadataFetcher.java]: an item -> items :)

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/InputData.java [hunk 11]
```diff
@@ -46,6 +46,7 @@
     protected String accessor;
     protected String resolver;
     protected String fragmenter;
+    protected String metadata;
```
Comment (hornn) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/InputData.java]: since this is actually the MetadataFetcher class name, perhaps rename metadataFetcher?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveDataFragmenter.java [hunk 12]
```diff
@@ -251,11 +251,11 @@ private void fetchTableMetaData(Metadata.Table tblDesc) throws Exception {
                 StorageDescriptor descPartition = partition.getSd();
                 props = MetaStoreUtils.getSchema(descPartition, descTable,
                         null, // Map<string, string> parameters - can be empty
-                        tblDesc.getDbName(), tblDesc.getTableName(), // table
+                        tblDesc.getPath(), tblDesc.getName(), // table
```
Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveDataFragmenter.java]: fix comment?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 13]
```diff
@@ -203,19 +205,37 @@ private static boolean verifyModifers(String[] modifiers) {
...
+        List<Metadata.Item> items = extractTablesFromPattern(null, qualifiedName);
+        if(items.isEmpty()) {
+            throw new IllegalArgumentException("No tables found");
+        }
+        return items.get(0);
+    }
 
-        String dbName, tableName;
+    /**
+     * Extracts the db_name and table_name from the qualifiedName.
```
Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: qualifiedName -> pattern?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 14]
```diff
@@ -203,19 +205,37 @@ private static boolean verifyModifers(String[] modifiers) {
...
+     * qualifiedName is the Hive table name or pattern that the user enters in the CREATE EXTERNAL TABLE statement
+     * or when querying HCatalog table.
+     * It can be either <code>table_name_pattern</code> or <code>db_name_pattern.table_name_pattern</code>.
+     *
+     * @param client Hivemetastore client
+     * @param pattern Hive table name or pattern
+     * @return {@link Metadata.Item} object holding the full table name
```
Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: returns list, right?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 15]
```diff
@@ -225,15 +245,47 @@ private static boolean verifyModifers(String[] modifiers) {
...
-            tableName = toks.get(1);
+            dbPattern = toks.get(0);
+            tablePattern = toks.get(1);
         } else {
-            throw new IllegalArgumentException("\"" + qualifiedName + "\"" + errorMsg);
+            throw new IllegalArgumentException("\"" + pattern + "\"" + errorMsg);
+        }
+
+        return getTablesFromPattern(client, dbPattern, tablePattern);
+
```
Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: remove extra line

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 16]
```diff
@@ -225,15 +245,47 @@ private static boolean verifyModifers(String[] modifiers) {
...
 
-        return new Metadata.Table(dbName, tableName);
+        try {
+            databases = client.getDatabases(dbPattern);
+            if(databases.isEmpty()) {
+                throw new IllegalArgumentException("No database found for the given pattern");
+            }
+            for(String dbName: databases) {
+                for(String tableName: client.getTables(dbName, tablePattern)) {
+                    if (!tableName.isEmpty()) {
```
Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: why do you need to check for an empty string?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 17]
```diff
@@ -225,15 +245,47 @@ private static boolean verifyModifers(String[] modifiers) {
...
         }
 
-        return new Metadata.Table(dbName, tableName);
+        try {
+            databases = client.getDatabases(dbPattern);
+            if(databases.isEmpty()) {
+                throw new IllegalArgumentException("No database found for the given pattern");
+            }
+            for(String dbName: databases) {
+                for(String tableName: client.getTables(dbName, tablePattern)) {
```
Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: can `getTables` accept as parameters both `dbName` pattern and `tablePattern` or does the `dbName` must be a real db?
If it's possible, perhaps we can get the list in one call.

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java [hunk 18]
```diff
@@ -28,7 +30,12 @@
  * abstract class which is returned by the MetadataFetcherFactory. 
  */
 public class MetadataFetcherFactory {
-    static public MetadataFetcher create(String fetcherName) throws Exception {
-        return (MetadataFetcher) Class.forName(fetcherName).newInstance();
+    /* TODO: This is a tempororary workaround.
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java]: typo temporary

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java [hunk 19]
```diff
@@ -28,7 +30,12 @@
...
+     * The metadata class will be moved to the pxf-profile.xml in the future
+     */
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java]: btw, why is it temporary? currently, both profile and specific plugin names are supported - so it will probably be possible to define a profile that includes the metadata fetcher, or specify the metadata fetcher itself.

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java [hunk 20]
```diff
@@ -55,50 +54,65 @@ public static String formatResponseString(Metadata metadata) throws IOException
      * To be used as the result string for HAWQ.
      * An example result is as follows:
      *
-     * {"PXFMetadata":[{"table":{"dbName":"default","tableName":"t1"},"fields":[{"name":"a","type":"int"},{"name":"b","type":"float"}]}]}
+     * {"PXFMetadata":[{"item":{"path":"default","name":"t1"},"fields":[{"name":"a","type":"int"},{"name":"b","type":"float"}]}]}
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: if it's a list of items, should there be more `[` or `{` in the jsob?

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java [hunk 21]
```diff
@@ -55,50 +54,65 @@ public static String formatResponseString(Metadata metadata) throws IOException
...
-        if ((metadata.getFields() == null) || metadata.getFields().isEmpty()) {
-            throw new IllegalArgumentException("metadata contains no fields - cannot serialize");
+        StringBuilder result = null;
+
+        for(Metadata metadata: metadataList) {
+            if(metadata == null) {
+                throw new IllegalArgumentException("metadata object is null - cannot serialize");
+            }
+            if ((metadata.getFields() == null) || metadata.getFields().isEmpty()) {
+                throw new IllegalArgumentException("metadata contains no fields - cannot serialize");
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: do you want to specify what is the faulty item or do we not care?

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/rest/MetadataResource.java [hunk 22]
```diff
@@ -56,51 +61,53 @@ public MetadataResource() throws IOException {
...
-                                                                                                                                     // nhorn
-                                                                                                                                     // -
-                                                                                                                                     // 09-03-15
-                                                                                                                                     // -
-                                                                                                                                     // pass
-                                                                                                                                     // as
-                                                                                                                                     // param
+            MetadataFetcher metadataFetcher = MetadataFetcherFactory.create(protData);
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/rest/MetadataResource.java]: aweome!

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java [hunk 23]
```diff
@@ -55,50 +54,65 @@ public static String formatResponseString(Metadata metadata) throws IOException
      * To be used as the result string for HAWQ.
      * An example result is as follows:
      *
-     * {"PXFMetadata":[{"table":{"dbName":"default","tableName":"t1"},"fields":[{"name":"a","type":"int"},{"name":"b","type":"float"}]}]}
+     * {"PXFMetadata":[{"item":{"path":"default","name":"t1"},"fields":[{"name":"a","type":"int"},{"name":"b","type":"float"}]}]}
      */
-    private static String metadataToJSON(Metadata metadata) throws IOException {
+    private static String metadataToJSON(List<Metadata> metadataList) throws IOException {
```
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: Have a look at the way we serialize the Fragments - we had a problem of having too many fragments, so we needed to serialize them inside a streaming object. 
I think that now that the metadata can include multiple tables, the same approach should be taken here as well.
The same applies also for the debug function that prints all of the items - if there are too many of them the StringBuilder will run out of memory. The solution in the fragments case was to print a log of one fragment at a time - perhaps the same should be done here as well.

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java (hunk 7 above)
Comment (hornn) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java]: perhaps call it MDFETCHER ?

Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/utilities/ProtocolData.java]: The first one i had in mind was DESIGNER :)

**Comment by hornn** [comment]
@shivzone Really nice changes!
Does it affect the way PXF interacts with HAWQ? IIRC we already assume the returned data will be a list of items, but I think the JSON field names are different now.
I think since the JSON answer is different, the API should change.

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java (hunk 1 above)
Comment (GodenYao) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/Metadata.java]: I meant to fix the error message: "Item name or Path cannot be empty"

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java (hunk 6 above)
Comment (GodenYao) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Also the grammar seems off :P , "Failed to connect to ..."

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java (hunk 19 above)
Comment (GodenYao) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java]: I think we've added this in profile xml. maybe the comment is obsolete.

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java [hunk 24]
```diff
@@ -55,50 +54,65 @@ public static String formatResponseString(Metadata metadata) throws IOException
...
-    private static String metadataToString(Metadata metadata) {
-        StringBuilder result = new StringBuilder("Metadata for table \"");
+    private static String metadataToString(List<Metadata> metadataList) {
+        StringBuilder result = new StringBuilder("Metadata:");
 
-        if (metadata == null) {
-            return "No metadata";
-        }
+        for(Metadata metadata: metadataList) {
+            result.append(" Metadata for item \"");
```
Comment (GodenYao) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: remove leading space.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java (hunk 6 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Updated case. Software engineers are not required to use perfect grammar :)

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java (hunk 16 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: will update, since we would anyway get a nosuchobjectexception if there are not tables

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java (hunk 17 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: getTables only takes absolute dbname and not dbpattern. so we will need 2 separate calls

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java (hunk 19 above)
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataFetcherFactory.java]: Updated

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java (hunk 20 above)
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: There can be, but the above example is still valid

**Comment by shivzone** [comment]
@sansanichfb is working on the related HAWQ changes to work with the new api

File: pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java (hunk 23 above)
Comment (shivzone) [inline pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/MetadataResponseFormatter.java]: Very valid point. Will handle this in a separate pull request/jira

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java [hunk 25]
```diff
@@ -203,19 +205,37 @@ private static boolean verifyModifers(String[] modifiers) {
...
      *
      * @param qualifiedName Hive table name
-     * @return {@link org.apache.hawq.pxf.api.Metadata.Table} object holding the full table name
+     * @return {@link Metadata.Item} object holding the full table name
      */
-    public static Metadata.Table parseTableQualifiedName(String qualifiedName) {
+    public static Metadata.Item extractTableFromName(String qualifiedName) {
+        List<Metadata.Item> items = extractTablesFromPattern(null, qualifiedName);
+        if(items.isEmpty()) {
+            throw new IllegalArgumentException("No tables found");
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: It's hard to parse stacktrace on client-side, so we should serialize all errors to JSON.

Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: We don't use JSON response for any of the other errors we throw for the other resources. We will have to be consistent across all the apis.

Comment (hornn) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: We have code in HAWQ (in libchurl) that already handles exceptions from PXF. It could be improved of course, but we cannot rely solely on json because some exceptions are thrown directly from tomcat, or are a result of connectivity issue.

Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: Yes, I saw that one, it's handle_special_error method in libchurl.c. We definitely should enhance error handling, but as for now the most common case is - when we didn't find any items matching specified pattern, so I think in this case we should just return empty response and 200 http code.

**Comment by sansanichfb** [comment]
+1, LGTM.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java (hunk 25 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilities.java]: I've updated to return default response "{"PXFMetadata":[]}" if no tables found
